subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/report/listener/ReportProcessorImpl.kt

```diff
@@ -52,85 +52,79 @@ internal class ReportProcessorImpl(
 
                         metricsSender.sendReportFileNotAvailable()
 
-                        val problem = createException(
+                        val problem = Problem(
                             shortDescription = errorMessage,
-                            context = "ReportProcessorImpl forms fallback Error status report",
+                            context = "ReportProcessor forms fallback Error status report",
                             because = "There is not enough context here about cause",
                             possibleSolutions = listOf(
                                 "MBS-11279 should help with correct error message from AdbDevice",
                                 "MBS-11281 to return tests with such errors back to retry queue"
                             ),
-                            cause = throwable
+                            throwable = throwable
                         )
 
                         processFailure(
-                            throwable = problem,
+                            problem = problem,
                             testStaticData = testFromSuite,
                             logcatBuffer = logcatBuffer
                         )
                     }.getOrThrow()
 
             is TestResult.Incomplete ->
                 with(result.infraError) {
-                    logger.warn("${error.message} while executing ${test.testName}", this.error)
+
+                    val problemBuilder = Problem.Builder(
+                        shortDescription = "Can't complete test execution",
+                        context = "ReportProcessor handling incomplete test result",
+                    )
+
+                    problemBuilder.throwable(error)
 
                     when (this) {
-                        is TestCaseRun.Result.Failed.InfrastructureError.FailedOnParsing ->
+                        is TestCaseRun.Result.Failed.InfrastructureError.FailedOnParsing -> {
                             metricsSender.sendFailedOnParsingInstrumentation()
+                            problemBuilder.because("Can't parse instrumentation output, see underlying exception")
+                        }
 
-                        is TestCaseRun.Result.Failed.InfrastructureError.FailedOnStart ->
+                        is TestCaseRun.Result.Failed.InfrastructureError.FailedOnStart -> {
                             metricsSender.sendFailedOnStartInstrumentation()
+                            problemBuilder.because("Can't start test")
+                        }
 
-                        is TestCaseRun.Result.Failed.InfrastructureError.Timeout ->
+                        is TestCaseRun.Result.Failed.InfrastructureError.Timeout -> {
                             metricsSender.sendTimeOut()
+                            problemBuilder.because(
+                                "Test didn't finish in time. " +
+                                    "Test Runner has hardcoded timeout of $timeoutMin minutes"
```

Comment:
> Test Runner has hardcoded timeout

This duplicates knowledge about implementation details. On this level, we shouldn't know it.

Reply:
I don't know how to deal with it better. 
On another level (where result is forming) i don't know about message detalization, it's just an exception message, not the same rich message for the end user like this `Problem`
Maybe it's a good idea to form problem on another level altogether

Reply:
> Maybe it's a good idea to form problem on another level altogether

Or make problems more composable.
It makes sense to append different pieces of knowledge incrementally on different layers of abstraction.

Maybe even compose `Prolem` itself to explicitly express the difference.

```
- Problem("short description for end-user", "possible solution on high level - disable the whole feature")
caused by:
- Problem("short description for devs", "possible solution on implementation level - support\change these implementation details; check something")
```
The point is to give pieces on different levels of understanding. The more I know, the more I can understand from it.

Reply:
It can be done. But it would be really hard to present such error. Like who want 40 lines of problem descriptions, even chained ones.

Reply:
Agree, sound reasonable